src/hotspot/os/windows/os_windows.cpp

```diff
@@ -867,21 +867,54 @@ int os::active_processor_count() {
     return ActiveProcessorCount;
   }
 
+  // Starting with Windows 11 and Windows Server 2022, the OS has changed to
+  // make processes and their threads span all processors in the system,
+  // across all processor groups, by default. Therefore, we will allow all
+  // processors to be active processors on these operating systems. However,
+  // job objects can be used to restrict processor affinity across the
+  // processor groups. In this case, the number of active processors must be
+  // obtained from the processor affinity in the job object.
+  bool schedules_all_processor_groups = win32::is_windows_11_or_greater() || win32::is_windows_server_2022_or_greater();
+  if (schedules_all_processor_groups) {
+    DWORD processors_in_job_object = win32::active_processors_in_job_object();
+
+    if (processors_in_job_object > 0) {
+      return processors_in_job_object;
+    }
+  }
+
+  DWORD logical_processors = 0;
+  SYSTEM_INFO si;
+  GetSystemInfo(&si);
+
+  // There is no associated job object so get the number of available logical processors from the process affinity mask
   DWORD_PTR lpProcessAffinityMask = 0;
   DWORD_PTR lpSystemAffinityMask = 0;
-  int proc_count = processor_count();
-  if (proc_count <= sizeof(UINT_PTR) * BitsPerByte &&
-      GetProcessAffinityMask(GetCurrentProcess(), &lpProcessAffinityMask, &lpSystemAffinityMask)) {
-    // Nof active processors is number of bits in process affinity mask
-    int bitcount = 0;
-    while (lpProcessAffinityMask != 0) {
-      lpProcessAffinityMask = lpProcessAffinityMask & (lpProcessAffinityMask-1);
-      bitcount++;
-    }
-    return bitcount;
+  if (GetProcessAffinityMask(GetCurrentProcess(), &lpProcessAffinityMask, &lpSystemAffinityMask)) {
+    logical_processors = win32::bit_count(lpProcessAffinityMask);
+
+    if (logical_processors != si.dwNumberOfProcessors) {
```

Comment:
Would it be clearer to make the test` if (logical_processors < si.dwNumberOfProcessors)`? I don't think it will make a difference to the behavior, but it's a clearer signal that affinity will result in a smaller number of processors to be active.